include/aws/common/linked_list.h

```diff
@@ -0,0 +1,91 @@
+#ifndef AWS_LINKED_LIST_H
+#define AWS_LINKED_LIST_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <stdint.h>
+
+struct aws_linked_list_node {
+    struct aws_linked_list_node *next;
+    struct aws_linked_list_node *prev;
+};
+
+struct aws_linked_list {
+    size_t length;
+    struct aws_linked_list_node *head;
+    struct aws_linked_list_node *tail;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+    * Initializes a doubly linked list.
+    */
+    AWS_COMMON_API void aws_linked_list_init(struct aws_linked_list *list);
+
+    /**
+    * Resets the list for either reuse or deletion.
+    */
+    AWS_COMMON_API void aws_linked_list_clean_up(struct aws_linked_list *list);
+
+    /**
+    * Pushes an element to the back of the list
+    */
+    AWS_COMMON_API void aws_linked_list_push_back(struct aws_linked_list *list, struct aws_linked_list_node *node);
+
+    /**
+    * Pushes an element to the front of the list.
+    */
+    AWS_COMMON_API void aws_linked_list_push_front(struct aws_linked_list *list, struct aws_linked_list_node *node);
+
+    /**
+    * Copies a memory address of the first element into item.
+    */
+    AWS_COMMON_API struct aws_linked_list_node* aws_linked_list_front(struct aws_linked_list *list);
```

Comment:
`head/tail` are more commonly how the head and tail of the list are referred to. Regardless, be consistent with naming of struct members vs naming of methods here.